Enhancement to the CLI: CaImportCertDirCommand.java

Stueypoo opened the pull request:
- Added an option to use an alternate CA certificate. 

When importing certificates into a newly migrated CA, the code performs a certificate verification based upon the CA's certificate that is stored in the DB. A problem with this is when the certificates being imported had been issued from the CA's previous certificate. An example CA that is impacted by this is a CSCA used for ePassports. These CA are renewed often, and the end-entity certificate have a long validity.

To remedy this, I added a "--cacert" option that allows the Operator to provide an alternate CA certificate that will then be used to verify the certificate being imported.

- Added an option to get the revocation details from the filename.

When importing revoked certificates, the Operator should supply the revocation REASON and INVALIDITY_TIME. If you have a lot of revoked certificates to import, then a better option is required. 

One idea is to have the REASON and TIME value encoded within the filename of each certificate, and the code then extracts these revocation details. The filename format being: CertName!REASON!TIME

  REASON is the reason code value or the name. eg., For suspended certs, use either "6" or "CERTIFICATEHOLD" or "CERTIFICATE_HOLD"
  TIME format is YYYY.MM.DD-hh.mm


## Checklist before requesting a review

- [X ] I have performed a self-review of my code
- [X ] I have kept the patch limited to only change the parts related to the patch
- [ ] This change requires a documentation update

primetomas commented:
Awesome, thank you. There is a system test for this command in modules/systemtests/src-test/org/ejbca/ui/cli/ca/CaImportCertCommandSystemTest.java

Can you add some testing of the new parameters there?

You can easily run the test with

`ant test:runone -Dtest.runone=CaImportCertCommandSystemTest`

Stueypoo commented:
Sure, I'll take a look at the system test cases.

Stueypoo commented:
Test cases added. Also updated the code to address the following:

- Change the filename seperator to be the exclamation mark (!)
- Revocation reason label can have underscore characters or not have them (IE., CERTIFICATE_HOLD or CERTIFICATEHOLD will work)
- The revocation time format of HH:mm is a problem with Windows filenames. Changed the format to HH.mm

System tests for this change are run with:
`ant test:runone -Dtest.runone=CaImportCertDirCommandTest`

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -174,7 +179,21 @@ public CommandResult execute(final ParameterContainer parameters) {
+            if ( caCertFile != null) {
+                final File fileCaCertFile = new File(caCertFile);
+                if (!fileCaCertFile.isFile()) {
+                    log.error("CA Certificate file '" + caCertFile + "' is not found. Will use current certificate for the CA.");
```
This should return a failure here:
return CommandResult.CLI_FAILURE;
since the arguments were not supplied properly, let the used know and fix the user error.

Stueypoo replied:
Yep, good point. Will fix.

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -115,6 +117,9 @@ public CommandResult execute(final ParameterContainer parameters) {
         final String revocationTimeString = parameters.get(REVOCATION_TIME);
         final int threadCount = parameters.get(THREAD_COUNT) == null ? 1 : Integer.valueOf(StringUtils.strip(parameters.get(THREAD_COUNT)));
+        final String caCertFile = parameters.get(CACERT);
+        final boolean bRevokeDetailsInFilename = parameters.containsKey(REVOKEDETAILS);
```
just call is revokedDetailsInFilename our coding convention does not use any type identifiers. Type is already clear.

Stueypoo replied:
OK. In fact, this boolean is only used the once, so might not really required.

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -1,19 +1,11 @@
-/*************************************************************************
- *                                                                       *
  *  EJBCA Community: The OpenSource Certificate Authority                *
```
The complete header must remain.

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -200,7 +219,7 @@ public CommandResult execute(final ParameterContainer parameters) {
                 return CommandResult.CLI_FAILURE;
             }
-
+ 
```
extra whitespace added, remove.

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -92,6 +89,11 @@ public class CaImportCertDirCommand extends BaseCaAdminCommand {
+        registerParameter(new Parameter(CACERT, "CA Certificate File", MandatoryMode.OPTIONAL, StandaloneMode.FORBID, ParameterMode.ARGUMENT,
+                "Specify an alternate CA certificate file (in PEM). Use this option when importing certificates that were issued by the previous CA certificate. Please note that the supplied certificate is not verified."));
+        registerParameter(new Parameter(REVOKEDETAILS, "Revocation Details", MandatoryMode.OPTIONAL, StandaloneMode.FORBID, ParameterMode.FLAG,
+                "Revocation details are to be derived from the filename of the certificate. The filename must end with '!<REASON>!<INVALIDITY_TIME>'. The REASON can be the value or label as described in RFC5280 section 5.3.1. INVALIDITY_TIME is formatted as '"+DATE_FORMAT_WINSAFE+"' and assumed to be the local timezone. Note: Filename extensions are not supported."));
```
what is "filename extensions"?.
It should be clear here that if revokedDetails flag is true, any filename that does not contain details will be ignored and not imported.

Stueypoo replied:
Will add more clarity as suggested.

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -214,6 +233,56 @@ public CommandResult execute(final ParameterContainer parameters) {
+                            final int iRevCode = Integer.parseInt(sRevCode);
+                            revocationReason = RevocationReasons.getFromDatabaseValue(iRevCode);
+                            if(revocationReason == null) {
+                                log.error("ERROR: '" + iRevCode + "' is not a valid revocation reason code. Ignoring this file.");
```
Message should print what file is ignored, otherwise it's impossible to know.

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -214,6 +233,56 @@ public CommandResult execute(final ParameterContainer parameters) {
+                                                             
+                            revocationReason = RevocationReasons.getFromCliValue(sRevCode.toUpperCase());
+                            if(revocationReason == null) {
+                                log.error("ERROR: '" + sRevCode + "' is not a valid revocation reason. Ignoring this file.");
```
Message should print what file is ignored, otherwise it's impossible to know.

Stueypoo replied:
OK

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -214,6 +233,56 @@ public CommandResult execute(final ParameterContainer parameters) {
+                       try {
+                            revocationTime = new SimpleDateFormat(DATE_FORMAT_WINSAFE).parse( sRevTime);
+                        } catch (ParseException e) {
+                            log.error("ERROR: '" + sRevTime + "' was not a valid revocation time. Use this time format '"+DATE_FORMAT_WINSAFE+"'. Ignoring this file");
```
Message should print what file is ignored, otherwise it's impossible to know.

primetomas commented on modules/systemtests/src-test/org/ejbca/ui/cli/ca/CaImportCertDirCommandTest.java:
```diff
@@ -1,19 +1,9 @@
-/*************************************************************************
- *                                                                       *
  *  EJBCA Community: The OpenSource Certificate Authority                *
```
add back header

Stueypoo replied:
Sorry about this. Not sure how this happened. Will fix.

primetomas commented on modules/ejbca-ejb-cli/src/org/ejbca/ui/cli/ca/CaImportCertDirCommand.java:
```diff
@@ -321,4 +390,11 @@ public String getFullHelpText() {
         return log;
     }
+    
+    private Certificate loadCertificateFromFile(final String filename) throws IOException, CertificateParsingException {
```
There exists a helper class to do this. CertTools.getCertsFromPEM((String certFilename, Class<T> returnType)). Use the helper class instead.

Stueypoo replied:
OK

primetomas reviewed (changes requested):
Some review comments. Nothing major.

Stueypoo commented:
Thanks @primetomas for the review.

primetomas commented:
Iäm sorry, but we just pushed EJBCA 9.0 to GitHub, which means the PR is messed up and have to be redone/rebased. 
Apart from that, I like it very much and I should be able to merge it next year.
Can you rebase the PR for the updated ejbca-ce git?

Stueypoo commented:
Closing and will re-open a new pull request based upon the latest main (v9) branch.
